src/data/releases/release.service.ts

```diff
@@ -19,57 +19,377 @@
+  let activeRelease = await getActiveReleaseOrThrow('No active release.');
+
+  const previousState = activeRelease.state;
+
+  if (isAllowedStateTransition(activeRelease, ReleaseState.PUBLISHING)) {
+    logger.info(
+      `Transitioning active release from ${previousState} to ${ReleaseState.PUBLISHING}.`,
+    );
+    const release = toPojo(
+      await releaseModel.updateRelease(activeRelease, {
+        state: ReleaseState.PUBLISHING,
+        error: null,
+      }),
+    );
+
+    return {
+      release,
+      previousState,
+      updated: true,
+      message: 'Publishing release...',
+    };
+  } else {
+    // Not allowed to make this state transition
+    logger.warn(
+      `Cannot transition active release from ${previousState} to ${ReleaseState.PUBLISHING}. No action taken.`,
+    );
+    return {
+      release: activeRelease,
+      previousState,
+      updated: false,
+      message:
+        previousState === ReleaseState.PUBLISHING
+          ? 'Already publishing release...'
+          : `Cannot begin publishing release that is in '${previousState}' state.`,
+    };
+  }
+}
+
+export async function finishPublishingActiveRelease(): Promise<ReleaseStateChangeResponse> {
+  const activeRelease = await getActiveReleaseOrThrow('No active release.');
+
+  const previousState = activeRelease.state;
+  if (isAllowedStateTransition(activeRelease, ReleaseState.PUBLISHED)) {
+    logger.info(`Transitioning active release from ${previousState} to ${ReleaseState.PUBLISHED}.`);
+    const release = toPojo(
+      await releaseModel.updateRelease(activeRelease, {
+        state: ReleaseState.PUBLISHED,
+        publishedAt: new Date(),
+      }),
+    );
+    return {
+      release,
+      previousState,
+      updated: true,
+      message: 'Active release finished publishing.',
+    };
+  } else {
+    return {
+      release: activeRelease,
+      previousState,
+      updated: false,
+      message: `Unable to finish publishing from '${previousState}' state.`,
+    };
+  }
+}
+
+/**
+ * Utility method for setActiveReleaseError
+ * @param previousState
+ * @returns
+ */
+function getErrorState(previousState: ReleaseState): ReleaseState | undefined {
+  switch (previousState) {
+    case ReleaseState.CALCULATING:
+      return ReleaseState.ERROR_CALCULATE;
+    case ReleaseState.BUILDING:
+      return ReleaseState.ERROR_BUILD;
+    case ReleaseState.PUBLISHING:
+      return ReleaseState.ERROR_PUBLISH;
+    default:
+      // No logic for handling an error from a static state.
+      return undefined;
+  }
+}
+export async function setActiveReleaseError(error: string) {
+  const activeRelease = await getActiveReleaseOrThrow('No active release.');
+
+  const previousState = activeRelease.state;
+  const nextState = getErrorState(previousState);
+  if (!nextState) {
+    throw new Error(
+      `Unable to set active release error state from '${previousState}'. Error message provided: ${error}`,
+    );
   }
```

Comment:
When an error occurs in processing, this allows us to put the release into the appropriate error state and save the error message in the release.